File: ios/Kurir/modules/login/LoginDelegate.swift
//
//  AuthDelegate.swift
//  Kurir
//
//  Created by Loren on 19/8/19.
//  Copyright © 2019 Crazybean Studio. All rights reserved.
//

import Foundation
import Crazybean
import Mobilex

class LoginDelegate: Delegate<LoginView, LoginViewModel> {
    func onLoginClick(name: String?, password: String?) {
        if let name = name, let password = password {
            view?.showSpinner()
            viewModel.login(name: name, password: password)
                .observe { [weak self] auth in
                    self?.view?.hideSpinner()
                    switch auth?.result {
                    case AuthKt.ERR_NONE:
                        self?.view?.showDashboard()
                        self?.view?.dismiss()
                        
                    case AuthKt.ERR_NOT_FOUND:
                        self?.view?.showNotFound()
                        
                    case AuthKt.ERR_PASSWORD:
                        self?.view?.showPasswordError()
                        
                    default:
                        self?.view?.dismiss()
                    }
            }
        }
    }
}

File: android/src/main/java/au/com/crazybean/mobilex/kurir/modules/login/LoginDelegate.kt
package au.com.crazybean.mobilex.kurir.modules.login

import androidx.lifecycle.Observer
import au.com.crazybean.foundation.mvvm.Delegate
import au.com.crazybean.mobilex.kurir.data.model.ERR_NONE
import au.com.crazybean.mobilex.kurir.data.model.ERR_NOT_FOUND
import au.com.crazybean.mobilex.kurir.data.model.ERR_PASSWORD

class LoginDelegate(view: LoginView?,
                    viewModel: LoginViewModel) : Delegate<LoginView, LoginViewModel>(view, viewModel) {

    fun onLoginClick(name: String?, password: String?) {
        when {
            name.isNullOrBlank() -> view?.showNameError()
            password.isNullOrBlank() -> view?.showPasswordError()
            else -> {
                view?.showSpinner()
                viewModel.login(name, password)
                    .observe(this, Observer { auth ->
                        view?.hideSpinner()
                        when (auth?.result) {
                            ERR_NONE -> {
                                view?.showDashboard()
                                view?.dismiss()
                            }
                            ERR_NOT_FOUND -> view?.showNotFound()
                            ERR_PASSWORD -> view?.showPasswordError()
                        }
                    })
            }
        }
    }

    fun onSignupClick() {
        view?.showRegister()
        view?.dismiss()
    }
}